opt: Project merged Upsert columns

andy-kimball opened the pull request:
Currently, the Upsert operator's insert and update columns are
separately created and maintained. For example:

  CREATE TABLE abc (a INT PRIMARY KEY, b INT)
  INSERT INTO abc VALUES (1) ON CONFLICT (a) DO UPDATE SET b=5

Here are the projected columns:

  SELECT
    a AS fetch_a, b AS fetch_b,
    1 AS ins_a, NULL AS ins_b,
    5 AS upd_b
  FROM ...

This commit constructs CASE statements that merge the insert and
update columns, like this:

  SELECT
    a AS fetch_a, b AS fetch_b,
    CASE WHEN fetch_a IS NULL THEN 1 ELSE fetch_a END AS ups_a,
    CASE WHEN fetch_a IS NULL THEN NULL::INT ELSE 5 END AS ups_b
  FROM ...

The advantages to making this change are:

1. This sets us up to implement check constraints, since they will now
   be able to operate on the combined columns.
2. This often avoids evaluating the update expression when an insert
   needs to occur for a given row.
3. This simplifies the property derivation, making Upsert act just like
   Update, Insert, and Delete. It's now possible to cleanly map each
   mutation output column to a single input column.

andy-kimball committed:
sql: Do not pass table descriptor by value

It is unnecessary to pass a table descriptor by value to TablesNeededForFKs.
This causes the entire struct to be copied by value onto the stack, and also
triggers a heap allocation when assigning the address of the stack struct
when setting the TableLookup.Table field.

Release note: None

andy-kimball committed:
opt: Wrap NULL values when desired type is not types.Any

When the type check phase returns a NULL value, wrap it with a cast
to the desired type (if it's not types.Any). Without this cast, the
expression's type is types.Unknown, which can cause typing issues
elsewhere in the code. In particular, it was causing a panic after
modifying the Upsert operator to use CASE expressions:

  CREATE TABLE ab (a INT, b INT[] AS (ARRAY[a]) STORED
  INSERT INTO ab (a) VALUES (NULL) ON CONFLICT (a) DO UPDATE SET b=ARRAY[1]

A projection like this is created and causes a panic:

  SELECT
    CASE WHEN canary IS NULL THEN ARRAY[NULL] ELSE ARRAY[1] END
  ...

The CASE operator is typed as UNKNOWN[], and DistSQL panics at runtime
when it encounters the INT[1] array.

The extra type cast ensures that we assign NULL values a static type that
will be propagated throughout the expression and prevent cases like this.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/33681)
<!-- Reviewable:end -->

RaduBerinde reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 9 of 9 files at r1, 8 of 10 files at r2, 11 of 11 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/33681#-:-LWCUFFL7JlssaSgC4bA:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/sql/opt/memo/expr.go, line 290 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/33681#-LWCSHMIAwUVDO2dhmCr:-LWCSHMJD2z3aBNt0S01:b3i5trl) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a4598cfdb4c677e51d0c1a2d9f496844761a680/pkg/sql/opt/memo/expr.go#L290)):*
> ```Go
> // (which holds the existing value of the column). And for the Upsert case, use
> // either the UpdateCols/FetchCols or the InsertCols value, as long as it is
> // non-zero. If both are non-zero, then they must be the same, since they are
> ```

[nit] maybe them being the same could be asserted inside the function? (unless it's performance concern)
___
*[pkg/sql/opt/optbuilder/insert.go, line 883 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/33681#-LWCTWtyCTVaOXQAVmm2:-LWCTWtyCTVaOXQAVmm3:b-27sfky) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a4598cfdb4c677e51d0c1a2d9f496844761a680/pkg/sql/opt/optbuilder/insert.go#L883)):*
> ```Go
> 		// column. This will be used by the Upsert operator in place of the
> 		// original columns.
> 		if mb.insertColList[i] != 0 {
> ```

[nit] Consider doing this inside each `case` above (it should simplify a lot in each case), I think it would be easier to follow what's going on.


<!-- Sent from Reviewable.io -->

andy-kimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/33681)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/sql/opt/memo/expr.go, line 290 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/33681#-LWCSHMIAwUVDO2dhmCr:-LWHDFj2At48O9lmbCzU:b-b4e8os) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a4598cfdb4c677e51d0c1a2d9f496844761a680/pkg/sql/opt/memo/expr.go#L290)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] maybe them being the same could be asserted inside the function? (unless it's performance concern)
</blockquote></details>

Just messier code, but I went ahead and made that change.
___
*[pkg/sql/opt/optbuilder/insert.go, line 883 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/33681#-LWCTWtyCTVaOXQAVmm2:-LWHDhqg53SFG6swWiP0:b-1w51rf) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a4598cfdb4c677e51d0c1a2d9f496844761a680/pkg/sql/opt/optbuilder/insert.go#L883)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] Consider doing this inside each `case` above (it should simplify a lot in each case), I think it would be easier to follow what's going on.
</blockquote></details>

I tried it, but I still have to do the `if` statement in case of `updateColList`, so I think it ends up being messier. I did update the comment a bit in hopes of making it clearer.


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
___
*[pkg/sql/opt/optbuilder/insert.go, line 883 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/33681#-LWCTWtyCTVaOXQAVmm2:-LWI21ZT2WQTP_Hndh0N:b-s1nmok) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a4598cfdb4c677e51d0c1a2d9f496844761a680/pkg/sql/opt/optbuilder/insert.go#L883)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I tried it, but I still have to do the `if` statement in case of `updateColList`, so I think it ends up being messier. I did update the comment a bit in hopes of making it clearer.
</blockquote></details>

I don't think you pushed your changes.


<!-- Sent from Reviewable.io -->

andy-kimball committed:
opt: Project merged Upsert columns

Currently, the Upsert operator's insert and update columns are
separately created and maintained. For example:

  CREATE TABLE abc (a INT PRIMARY KEY, b INT)
  INSERT INTO abc VALUES (1) ON CONFLICT (a) DO UPDATE SET b=5

Here are the projected columns:

  SELECT
    a AS fetch_a, b AS fetch_b,
    1 AS ins_a, NULL AS ins_b,
    5 AS upd_b
  FROM ...

This commit constructs CASE statements that merge the insert and
update columns, like this:

  SELECT
    a AS fetch_a, b AS fetch_b,
    CASE WHEN fetch_a IS NULL THEN 1 ELSE fetch_a END AS ups_a,
    CASE WHEN fetch_a IS NULL THEN NULL::INT ELSE 5 END AS ups_b
  FROM ...

The advantages to making this change are:

1. This sets us up to implement check constraints, since they will now
   be able to operate on the combined columns.
2. This often avoids evaluating the update expression when an insert
   needs to occur for a given row.
3. This simplifies the property derivation, making Upsert act just like
   Update, Insert, and Delete. It's now possible to cleanly map each
   mutation output column to a single input column.

Release note: None

andy-kimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/33681)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/sql/opt/optbuilder/insert.go, line 883 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/33681#-LWCTWtyCTVaOXQAVmm2:-LWMzm_N5S_agbUOQxw3:bo4gwcf) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a4598cfdb4c677e51d0c1a2d9f496844761a680/pkg/sql/opt/optbuilder/insert.go#L883)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I don't think you pushed your changes.
</blockquote></details>

I ended up finding a bug once I added the assert, and had to change a few things. Should be correct now.


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 5 of 24 files at r4, 8 of 10 files at r5, 13 of 13 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/33681#-:-LWN1ILS0-cDI0eUHOLV:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained

___
*[pkg/sql/opt/optbuilder/insert.go, line 883 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/33681#-LWCTWtyCTVaOXQAVmm2:-LWN0vhl1tz2sp3zax-K:b-83sjoh) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a4598cfdb4c677e51d0c1a2d9f496844761a680/pkg/sql/opt/optbuilder/insert.go#L883)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I ended up finding a bug once I added the assert, and had to change a few things. Should be correct now.
</blockquote></details>

Nice!


<!-- Sent from Reviewable.io -->

andy-kimball commented:
bors r+

andy-kimball commented:
bors r+

andy-kimball commented:
bors r+
